### PR TITLE
Add Multilayer Perceptron Regression

### PR DESCRIPTION
**Pull request by JeremyNixon**
## What changes were proposed in this pull request?

This is a pull request adding support for Multilayer Perceptron Regression, the counterpart to the Multilayer Perceptron Classifier (hereafter MLPR and MLPC). 
#### Outline
1. Major Changes
2. API Decisions
3. Automating Scaling
4. Features
5. Reference Resources
## Major Changes

There are two major differences between MLPR and MLPC. The first is the use of an linear (identity) activation function and a sum of squared error cost function in the last layer of the network. The second is the requirement to scale the data to [0,1] and back to make it easy for the weights to fit a value in the proper range. 
#### Linear, Relu, Tanh Activations

In the forward pass the linear activation passes the value from the fully connected layer through to become the network prediction. In weight adjustment during the backward pass its derivative is one. All regression models will use the linear activation in the last layer, and so there is no option (as there is in MLPC) to use another activation function and cost function in the last layer.

The Relu and Tanh are activation functions that will benefit the accuracy and convergence speed MLPC and MLPR. Tanh zero-centers the data passed to neurons which aids optimization. Relu avoids saturating the gradients.
#### Automated Scaling

The data scaling is done through min-max scaling, where the minimum label is subtracted from every value (leading to a range from [0 to max-min]) and then dividing by max-min to get a scale from 0 to 1. The corner case where max-min = 0 is resolved by omitting the division step. 
#### Motivating Example

screen shot 2016-09-29 at 4 17 23 pm
## API Decisions

The API is identical to MLPC with the exception of softmaxOnTop - there is no option on the last layer activation function, or on the cost function to be used (MLPC gives a choice between cross entropy and sum of square error). This API has the user call MLPR with a set of layers that represent the topology of the network. The number of hidden layers is inferred by the parameter for the labels and is equal to the total number of layers - 2. Each hidden layer will be a feedforward layer with a sigmoid activation function up to the output layer and its linear activation.
#### Input/Output Layer Argument

For MLPR, the output count will usually be 1, and the number of inputs will always be equal to the number of features in the training dataset. One API choice could be to omit the input and output counts and only have the user supply the number of neurons in the hidden layers, and automate the input and output counts by looking at the training data. At the very least, it makes sense to validate the user’s layer parameter and display a helpful error message instead of the error in the data stacker that currently appears if the improper number of inputs or outputs is provided.
#### Modular API

It also would make sense for the API to be modular. A user will want the flexibility to use the linear layer at different points in the network (as well as in MLPC), and will certainly want to be able to use new activation functions (tanh, relu) that are added to improve the performance of these models. That flexibility allows a user to tune the network to their dataset and will be particularly important for convnets or recurrent nets in future. We should decide on the best way to enable tanh and relu activations in this algorithm and in the classifier for the time being.
## Automating Scaling

Current behavior has an argument allowing the user to turn off scaling using MultilayerPerceptronRegressor().setStandardizeLabels(false).

In general I advocate for helpful defaults that can be overridden, where we scale automatically but give an option to run without scaling and don’t run autoscaling if both the min and max are provided by the user.
## Features

There are a few features that have been checked:
1. Integrates cleanly with pipeline API
2. Model save/load is enabled
## Reference Resources

Christopher M. Bishop. Neural Networks for Pattern Recognition.
Patrick Nicolas. Scala for Machine Learning, Chapter 9.
Ian Goodfellow Yoshua Bengio and Aaron Courville. Deep Learning, Chapter 6.

### CODE REVIEW

**Comment by JeremyNixon** [comment]
I would like to add a contribution guide as well - contributors need to clone the repo and publish to local if they want to iterate over their changes. Specifically, they need to change the version and use either sbt publish-local or sbt publishM2, and then point spark to the changed version of the package. This PR doesn't include a test suite / example - I plan to add the test suite, and would like to know if you'd prefer the example come with internal data or not.

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Dataset
+import org.apache.spark.sql.functions.{max, min}
+import org.apache.spark.sql.types._
+
+import scaladl.layers.{FeedForwardTopology, FeedForwardTrainer}
+
+
+/** Params for Multilayer Perceptron. */
+private[ml] trait MultilayerPerceptronBasicParams extends PredictorParams
```
Comment (avulanov) [inline src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: Could you use existing `MultilayerPerceptronParams`?

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+  final val minimum: DoubleParam = new DoubleParam(this, "min",
+    "Minimum value for scaling data.")
+
+  /**
+   * Set the minimum value in the training set labels.
+   *
+   * @group setParam
+   */
+  def setMin(value: Double): this.type = set(minimum, value)
```
Comment (avulanov) [inline src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: Do we need to set it? It can be extracted from the training set automatically.

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+  final val maximum: DoubleParam = new DoubleParam(this, "max",
+    "Max value for scaling data.")
+
+  /**
+   * Set the maximum value in the training set labels.
+   *
+   * @group setParam
+   */
+  def setMax(value: Double): this.type = set(maximum, value)
```
Comment (avulanov) [inline src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: Ditto

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+  override def copy(extra: ParamMap): MultilayerPerceptronRegressor = defaultCopy(extra)
+
+  /**
+   * Train a model using the given dataset and parameters.
+   *
+   * @param dataset Training dataset
+   * @return Fitted model
+   */
+  override protected def train(dataset: Dataset[_]): MultilayerPerceptronRegressorModel = {
```
Comment (avulanov) [inline src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala:181]: We should support multiple outputs, i.e. `dataset` should contain two vector columns - `input` and `output`.

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    if (getStandardizeLabels) {
+      // Compute minimum and maximum values in the training labels for scaling.
+      val minmax = dataset
+        .agg(max("label").cast(DoubleType), min("label").cast(DoubleType)).collect()(0)
+      setMin(minmax(1).asInstanceOf[Double])
+      setMax(minmax(0).asInstanceOf[Double])
+    }
+    // Encode and scale labels to prepare for training.
+    val data = lpData.map(lp =>
+      RegressionLabelConverter.encodeLabeledPoint(lp, $(minimum), $(maximum)))
```
Comment (avulanov) [inline src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: Not needed if multiple outputs are supported

File: src/main/scala/scaladl/layers/Layer.scala [hunk 6]
```diff
@@ -384,6 +423,26 @@ object FeedForwardTopology {
...
+   * @param layerSizes sizes of layers including input and output size
+   * @return multilayer perceptron topology
+   */
+  def multiLayerPerceptronRegression(layerSizes: Array[Int]): FeedForwardTopology = {
+    val layers = new Array[Layer]((layerSizes.length - 1) * 2)
+    for (i <- 0 until layerSizes.length - 1) {
+      layers(i * 2) = new AffineLayer(layerSizes(i), layerSizes(i + 1))
+      layers(i * 2 + 1) =
+        if (i == layerSizes.length - 2) {
+          new LinearLayerWithSquaredError()
```
Comment (avulanov) [inline src/main/scala/scaladl/layers/Layer.scala]: Could you reuse `EmptyLayerWithSquaredError`? Later we can refactor it to be consistent with the other loss layers.

File: src/main/scala/scaladl/layers/Layer.scala [hunk 7]
```diff
@@ -384,6 +423,26 @@ object FeedForwardTopology {
...
+        } else {
+          new FunctionalLayer(new TanhFunction())
```
Comment (avulanov) [inline src/main/scala/scaladl/layers/Layer.scala]: Lets use `new FunctionalLayer(new SigmoidFunction())` for this PR and add other layer types in a separate PR.

**Review by avulanov — COMMENTED** [review]
@JeremyNixon Thank you! I have done a first pass.

File: src/main/scala/scaladl/layers/Layer.scala (hunk 6 above)
Comment (mhaut) [reply src/main/scala/scaladl/layers/Layer.scala]: Hi @avulanov and @JeremyNixon in order to refactor code I think that is a good idea add a typelayer to function:

```
val layers = Array[Int](2, 5, 2)

val layerstype = Array[String]("sigmoid", "softmaxOnTop")

val trainer = new MultilayerPerceptronClassifier()
```

If you can try, you can clone this repo: https://github.com/mhaut/scalable-deeplearning
I try to pull request also.

Comment (avulanov) [reply src/main/scala/scaladl/layers/Layer.scala]: @mhaut Sounds good. However it needs more than that to be generic. The name of the layer should be stored within Layer itself. There should be a layer fabric that can instantiate layers based on their names and set up the parameters. This is just high-level description, the actual design should be though trough.

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala (hunk 1 above)
Comment (JeremyNixon) [reply src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: Absolutely can. I think that MultilayerPerceptronParams is generic, and so with both algorithms is seems odd to have it in the classifier. What do you think of moving it to Layer.scala or another more general file?

File: src/main/scala/scaladl/layers/Layer.scala (hunk 7 above)
Comment (JeremyNixon) [reply src/main/scala/scaladl/layers/Layer.scala]: Makes sense.

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala (hunk 2 above)
Comment (JeremyNixon) [reply src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: We don't need to set it - I updated this to extract it without using a param.

File: src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala (hunk 3 above)
Comment (JeremyNixon) [reply src/main/scala/org/apache/spark/ml/scaladl/MultilayerPerceptronRegressor.scala]: Ditto
